File: backend/app/schemas/repository.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class LanguageBreakdown(BaseModel):
    """A single language's share of the repository, by bytes of code."""

    name: str
    bytes: int
    percentage: float = Field(..., description="Rounded percentage of total bytes, 0-100.")


class RepositoryOwner(BaseModel):
    login: str
    avatar_url: str
    html_url: str
    type: str


class RepositoryOverview(BaseModel):
    """Normalized repository metadata — the application's own shape, not GitHub's raw response."""

    owner: RepositoryOwner
    name: str
    full_name: str
    description: str | None = None
    html_url: str
    homepage: str | None = None

    stars: int
    forks: int
    watchers: int
    open_issues: int

    default_branch: str
    primary_language: str | None = None
    topics: list[str] = Field(default_factory=list)

    is_archived: bool
    is_fork: bool
    license_name: str | None = None

    created_at: datetime
    updated_at: datetime
    pushed_at: datetime | None = None

    languages: list[LanguageBreakdown] = Field(default_factory=list)


class RepositoryLanguagesResponse(BaseModel):
    languages: list[LanguageBreakdown]

File: frontend/types/repository.ts
export interface RepositoryOwner {
  login: string;
  avatar_url: string;
  html_url: string;
  type: string;
}

export interface LanguageBreakdown {
  name: string;
  bytes: number;
  percentage: number;
}

export interface RepositoryOverview {
  owner: RepositoryOwner;
  name: string;
  full_name: string;
  description: string | null;
  html_url: string;
  homepage: string | null;

  stars: number;
  forks: number;
  watchers: number;
  open_issues: number;

  default_branch: string;
  primary_language: string | null;
  topics: string[];

  is_archived: boolean;
  is_fork: boolean;
  license_name: string | null;

  created_at: string;
  updated_at: string;
  pushed_at: string | null;

  languages: LanguageBreakdown[];
}

export interface ApiErrorBody {
  error: string;
  message: string;
  detail?: string;
}

export class ApiError extends Error {
  status: number;
  code: string;
  detail?: string;

  constructor(status: number, body: ApiErrorBody) {
    super(body.message);
    this.status = status;
    this.code = body.error;
    this.detail = body.detail;
  }
}
